### ISSUE TITLE
make fails

### ISSUE THREAD

**Issue by David-Lor** [Overv/vramfs#17]
1. Output of `make`:

```bash
david@Ubuntu-Main:~/vramfs$ make
g++ -Wall -Wpedantic -Werror -std=c++11 -D_FILE_OFFSET_BITS=64 -I/usr/include/fuse -I include/ -march=native -O2 -flto -c -o build/util.o src/util.cpp
g++ -Wall -Wpedantic -Werror -std=c++11 -D_FILE_OFFSET_BITS=64 -I/usr/include/fuse -I include/ -march=native -O2 -flto -c -o build/memory.o src/memory.cpp
In file included from src/memory.cpp:1:0:
include/memory.hpp:14:14: fatal error: CL/cl2.hpp: No existe el archivo o el directorio
     #include <CL/cl2.hpp>
```

2. Output of `make DEBUG=1`:

```bash
david@Ubuntu-Main:~/vramfs$ make DEBUG=1
g++ -Wall -Wpedantic -Werror -std=c++11 -D_FILE_OFFSET_BITS=64 -I/usr/include/fuse -I include/ -g -DDEBUG -Wall -Werror -std=c++11 -c -o build/memory.o src/memory.cpp
src/memory.cpp: In function ‘int vram::memory::clear_buffer(const cl::Buffer&)’:
src/memory.cpp:24:61: error: binding reference of type ‘cl::Buffer&’ to ‘const cl::Buffer’ discards qualifiers
                 return queue.enqueueCopyBuffer(zero_buffer, buf, 0, 0, block::size, nullptr, nullptr);
                                                             ^~~
In file included from include/memory.hpp:12:0,
                 from src/memory.cpp:1:
include/CL/debugcl.hpp:80:13: note:   initializing argument 2 of ‘int cl::CommandQueue::enqueueCopyBuffer(const cl::Buffer&, cl::Buffer&, int, int, int, std::vector<cl::Event>*, cl::Event*)’
         int enqueueCopyBuffer(const Buffer& src, Buffer& dst, int offSrc, int offDst, int size, std::vector<cl::Event>* events, cl::Event* event) {
             ^~~~~~~~~~~~~~~~~
src/memory.cpp: In function ‘bool vram::memory::init_opencl()’:
src/memory.cpp:44:17: error: ‘cl_uint’ was not declared in this scope
                 cl_uint version = cl::detail::getPlatformVersion(platform());
                 ^~~~~~~
src/memory.cpp:44:17: note: suggested alternative: ‘cl_int’
                 cl_uint version = cl::detail::getPlatformVersion(platform());
                 ^~~~~~~
                 cl_int
src/memory.cpp:46:21: error: ‘version’ was not declared in this scope
                 if (version >= (1 << 16 | 2))
                     ^~~~~~~
src/memory.cpp:52:55: error: ‘CL_MEM_READ_ONLY’ was not declared in this scope
                     zero_buffer = cl::Buffer(context, CL_MEM_READ_ONLY | CL_MEM_COPY_HOST_PTR, block::size, zero_data, &r);
                                                       ^~~~~~~~~~~~~~~~
src/memory.cpp:52:55: note: suggested alternative: ‘CL_MEM_READ_WRITE’
                     zero_buffer = cl::Buffer(context, CL_MEM_READ_ONLY | CL_MEM_COPY_HOST_PTR, block::size, zero_data, &r);
                                                       ^~~~~~~~~~~~~~~~
                                                       CL_MEM_READ_WRITE
src/memory.cpp:52:74: error: ‘CL_MEM_COPY_HOST_PTR’ was not declared in this scope
                     zero_buffer = cl::Buffer(context, CL_MEM_READ_ONLY | CL_MEM_COPY_HOST_PTR, block::size, zero_data, &r);
                                                                          ^~~~~~~~~~~~~~~~~~~~
src/memory.cpp: In function ‘std::vector<std::__cxx11::basic_string<char> > vram::memory::list_devices()’:
src/memory.cpp:87:51: error: ‘class cl::Device’ has no member named ‘getInfo’
                     device_names.push_back(device.getInfo<CL_DEVICE_NAME>());
                                                   ^~~~~~~
src/memory.cpp:87:59: error: ‘CL_DEVICE_NAME’ was not declared in this scope
                     device_names.push_back(device.getInfo<CL_DEVICE_NAME>());
                                                           ^~~~~~~~~~~~~~
src/memory.cpp:87:59: note: suggested alternative: ‘CL_DEVICE_TYPE_GPU’
                     device_names.push_back(device.getInfo<CL_DEVICE_NAME>());
                                                           ^~~~~~~~~~~~~~
                                                           CL_DEVICE_TYPE_GPU
src/memory.cpp:87:75: error: expected primary-expression before ‘)’ token
                     device_names.push_back(device.getInfo<CL_DEVICE_NAME>());
                                                                           ^
Makefile:18: recipe for target 'build/memory.o' failed
make: *** [build/memory.o] Error 1
```

3. Installed `ocl-icd-opencl-dev` (`2.2.11-1ubuntu1`) & `libfuse-dev` (`2.9.7-1ubuntu1`)

4. OS: Ubuntu 18.04

5. GPU: GTX 970

6. GPU drivers: Nvidia 390.116

7. Output of `clinfo`:

```bash
david@Ubuntu-Main:~/vramfs$ clinfo
Number of platforms                               1
  Platform Name                                   NVIDIA CUDA
  Platform Vendor                                 NVIDIA Corporation
  Platform Version                                OpenCL 1.2 CUDA 9.1.84
  Platform Profile                                FULL_PROFILE
  Platform Extensions                             cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_fp64 cl_khr_byte_addressable_store cl_khr_icd cl_khr_gl_sharing cl_nv_compiler_options cl_nv_device_attribute_query cl_nv_pragma_unroll cl_nv_copy_opts cl_nv_create_buffer
  Platform Extensions function suffix             NV

  Platform Name                                   NVIDIA CUDA
Number of devices                                 1
  Device Name                                     GeForce GTX 970
  Device Vendor                                   NVIDIA Corporation
  Device Vendor ID                                0x10de
  Device Version                                  OpenCL 1.2 CUDA
  Driver Version                                  390.116
  Device OpenCL C Version                         OpenCL C 1.2 
  Device Type                                     GPU
  Device Topology (NV)                            PCI-E, 01:00.0
  Device Profile                                  FULL_PROFILE
  Device Available                                Yes
  Compiler Available                              Yes
  Linker Available                                Yes
  Max compute units                               13
  Max clock frequency                             1329MHz
  Compute Capability (NV)                         5.2
  Device Partition                                (core)
    Max number of sub-devices                     1
    Supported partition types                     None
  Max work item dimensions                        3
  Max work item sizes                             1024x1024x64
  Max work group size                             1024
  Preferred work group size multiple              32
  Warp size (NV)                                  32
  Preferred / native vector sizes                 
    char                                                 1 / 1       
    short                                                1 / 1       
    int                                                  1 / 1       
    long                                                 1 / 1       
    half                                                 0 / 0        (n/a)
    float                                                1 / 1       
    double                                               1 / 1        (cl_khr_fp64)
  Half-precision Floating-point support           (n/a)
  Single-precision Floating-point support         (core)
    Denormals                                     Yes
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 Yes
    Round to infinity                             Yes
    IEEE754-2008 fused multiply-add               Yes
    Support is emulated in software               No
    Correctly-rounded divide and sqrt operations  Yes
  Double-precision Floating-point support         (cl_khr_fp64)
    Denormals                                     Yes
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 Yes
    Round to infinity                             Yes
    IEEE754-2008 fused multiply-add               Yes
    Support is emulated in software               No
  Address bits                                    64, Little-Endian
  Global memory size                              4234018816 (3.943GiB)
  Error Correction support                        No
  Max memory allocation                           1058504704 (1009MiB)
  Unified memory for Host and Device              No
  Integrated memory (NV)                          No
  Minimum alignment for any data type             128 bytes
  Alignment of base address                       4096 bits (512 bytes)
  Global Memory cache type                        Read/Write
  Global Memory cache size                        212992 (208KiB)
  Global Memory cache line size                   128 bytes
  Image support                                   Yes
    Max number of samplers per kernel             32
    Max size for 1D images from buffer            134217728 pixels
    Max 1D or 2D image array size                 2048 images
    Max 2D image size                             16384x16384 pixels
    Max 3D image size                             4096x4096x4096 pixels
    Max number of read image args                 256
    Max number of write image args                16
  Local memory type                               Local
  Local memory size                               49152 (48KiB)
  Registers per block (NV)                        65536
  Max number of constant args                     9
  Max constant buffer size                        65536 (64KiB)
  Max size of kernel argument                     4352 (4.25KiB)
  Queue properties                                
    Out-of-order execution                        Yes
    Profiling                                     Yes
  Prefer user sync for interop                    No
  Profiling timer resolution                      1000ns
  Execution capabilities                          
    Run OpenCL kernels                            Yes
    Run native kernels                            No
    Kernel execution timeout (NV)                 Yes
  Concurrent copy and kernel execution (NV)       Yes
    Number of async copy engines                  2
  printf() buffer size                            1048576 (1024KiB)
  Built-in kernels                                
  Device Extensions                               cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_fp64 cl_khr_byte_addressable_store cl_khr_icd cl_khr_gl_sharing cl_nv_compiler_options cl_nv_device_attribute_query cl_nv_pragma_unroll cl_nv_copy_opts cl_nv_create_buffer

NULL platform behavior
  clGetPlatformInfo(NULL, CL_PLATFORM_NAME, ...)  NVIDIA CUDA
  clGetDeviceIDs(NULL, CL_DEVICE_TYPE_ALL, ...)   Success [NV]
  clCreateContext(NULL, ...) [default]            Success [NV]
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_DEFAULT)  No platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_CPU)  No devices found in platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_GPU)  No platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_ACCELERATOR)  No devices found in platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_CUSTOM)  Invalid device type for platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_ALL)  No platform

ICD loader properties
  ICD loader Name                                 OpenCL ICD Loader
  ICD loader Vendor                               OCL Icd free software
  ICD loader Version                              2.2.11
  ICD loader Profile                              OpenCL 2.1
```

**Comment by Overv** [comment]
You are probably missing the `opencl-headers` package on Ubuntu, because it works fine for me on Arch. You are right about the debug build being broken, though. I've just fixed that in e2369ed24e07b893beb5ccf755986594af2e8651.

**Comment by David-Lor** [comment]
Indeed, after installing `opencl-headers`, I was able to run make and run it. Working now, thanks.
